### PR TITLE
Rest api muc light config

### PR DESCRIPTION
**Pull request by zofpolkowska**
Add possibility of changing muc light room configuration through REST API

### CODE REVIEW

**Comment by mongoose-im** [comment]
[5376.1](https://travis-ci.org/esl/MongooseIM/jobs/415864607) / Erlang 19.3 / small_tests / 0180f270db10831a4b7281fa2bf33d374337c091
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5376/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5376/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-14_11.20.53/lib.mongooseim.logs/run.2018-08-14_11.20.54/suite.log.html)

**Comment by mongoose-im** [comment]
[5377.1](https://travis-ci.org/esl/MongooseIM/jobs/415970809) / Erlang 19.3 / small_tests / c6bd30c63dd35a06003636f88d57214e1bc5be5b
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5377/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5377/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-14_15.33.03/lib.mongooseim.logs/run.2018-08-14_15.33.04/suite.log.html)

**Comment by mongoose-im** [comment]
[5382.1](https://travis-ci.org/esl/MongooseIM/jobs/418109346) / Erlang 19.3 / small_tests / 985a5d1c18350bca74912a587a5e9068adff5b53
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5382/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5382/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-20_07.43.18/lib.mongooseim.logs/run.2018-08-20_07.43.19/suite.log.html)

**Comment by mongoose-im** [comment]
[5384.1](https://travis-ci.org/esl/MongooseIM/jobs/418169309) / Erlang 19.3 / small_tests / d2c3953824db5fae4e5f30e74e9dacf6c192f6ea
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5384/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5384/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-20_10.49.44/lib.mongooseim.logs/run.2018-08-20_10.49.44/suite.log.html)

**Comment by mongoose-im** [comment]
[5387.1](https://travis-ci.org/esl/MongooseIM/jobs/418179228) / Erlang 19.3 / small_tests / 3334ec5dcaaf6397bb04071341bdb575214f78fb
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5387/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5387/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-20_11.24.04/lib.mongooseim.logs/run.2018-08-20_11.24.05/suite.log.html)

File: big_tests/tests/rest_client_SUITE.erl [hunk 1]
```diff
@@ -35,7 +35,8 @@ all() ->
 groups() ->
     G = [{messages_with_props, [parallel], message_with_props_test_cases()},
          {messages, [parallel], message_test_cases()},
-         {muc, [parallel], muc_test_cases()},
+         {muc,[pararell] , muc_test_cases()},
```
Comment (kzemek) [inline big_tests/tests/rest_client_SUITE.erl]: Slight misformat

**Comment by mongoose-im** [comment]
[5393.1](https://travis-ci.org/esl/MongooseIM/jobs/418253356) / Erlang 19.3 / small_tests / e61b4eff2da108222a68bf2b39d2021edffde3cf
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5393/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5393/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-20_14.32.12/lib.mongooseim.logs/run.2018-08-20_14.32.13/suite.log.html)

File: big_tests/tests/rest_client_SUITE.erl [hunk 2]
```diff
@@ -127,6 +145,15 @@ init_per_testcase(TC, Config) ->
                    ],
     rest_helper:maybe_skip_mam_test_cases(TC, MAMTestCases, Config).
 
+end_per_testcase(config_can_be_changed_by_all = CaseName, Config) ->
+    Host = ct:get_config({hosts, mim, domain}),
+    MUCLightHost = <<"muclight.", Host/binary>>,
+    dynamic_modules:restart(Host, mod_muc_light,
```
Comment (kzemek) [inline big_tests/tests/rest_client_SUITE.erl]: A less error-prone way would be to call `dynamic_modules:save_modules/2` before the testcase and `dynamic_modules:restore_modules/2` after. See amp_big_SUITE.erl for example

File: big_tests/tests/rest_client_SUITE.erl [hunk 3]
```diff
@@ -251,6 +278,45 @@ room_is_created_with_given_identifier(Config) ->
         assert_room_info(Alice, RoomInfo)
     end).
 
+config_can_be_changed_by_owner(Config) ->
+    escalus:fresh_story(Config, [{alice, 1}], fun(Alice) ->
+        RoomID = <<"som_othere_id">>,
+        RoomJID = room_jid(RoomID, Config),
+        RoomID = given_new_room({alice, Alice}, RoomJID, <<"old_name">>),
+        RoomInfo = get_room_info({alice, Alice}, RoomID),
+        assert_property_value(<<"name">>,<<"old_name">>,RoomInfo),
```
Comment (kzemek) [inline big_tests/tests/rest_client_SUITE.erl]: Missing whitespace between args

File: big_tests/tests/rest_client_SUITE.erl [hunk 4]
```diff
@@ -251,6 +278,45 @@ room_is_created_with_given_identifier(Config) ->
...
+
+        {{<<"200">>, <<"OK">>}, {_}} =
+            given_config_change({alice, Alice}, RoomJID, <<"new_name">>, <<"new_subject">>),
+        NewRoomInfo = get_room_info({alice, Alice}, RoomID),
+        assert_property_value(<<"name">>, <<"new_name">>, NewRoomInfo),
+        assert_property_value(<<"subject">>, <<"new_subject">>, NewRoomInfo)
+    end).
+
+config_cannot_be_changed(Config) ->
```
Comment (kzemek) [inline big_tests/tests/rest_client_SUITE.erl]: The test should have at least a single word in the name indicating *why* the config cannot be changed.

File: big_tests/tests/rest_client_SUITE.erl [hunk 5]
```diff
@@ -734,6 +810,15 @@ is_property_present(Name, Proplist) ->
     Val = proplists:get_value(Name, Proplist),
     Val /= undefined.
 
+assert_property_value(Name, Value, Proplist) ->
+    Val = proplists:get_value(Name, Proplist),
+    case Val of
+        Value ->
+            true;
+        _ ->
+            ct:fail(#{issue => config_change_failed})
```
Comment (kzemek) [inline big_tests/tests/rest_client_SUITE.erl]: In terms of info given on failure, this seems to be slightly worse than a simple `?assertEqual(Value, Val)`.

**Comment by mongoose-im** [comment]
[5396.1](https://travis-ci.org/esl/MongooseIM/jobs/418300053) / Erlang 19.3 / small_tests / 980ecd5c97286a94023a72c8628f851770995e16
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5396/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5396/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-20_16.08.36/lib.mongooseim.logs/run.2018-08-20_16.08.37/suite.log.html)

**Comment by mongoose-im** [comment]
[5398.1](https://travis-ci.org/esl/MongooseIM/jobs/418309195) / Erlang 19.3 / small_tests / 175b0de399d771737e0d7e347d80154838f7f6d3
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5398/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5398/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-20_17.00.08/lib.mongooseim.logs/run.2018-08-20_17.00.09/suite.log.html)

**Comment by mongoose-im** [comment]
[5401.1](https://travis-ci.org/esl/MongooseIM/jobs/418397395) / Erlang 19.3 / small_tests / 9b855116166f76df6e70263fcec572891b55f417
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5401/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5401/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-20_20.17.38/lib.mongooseim.logs/run.2018-08-20_20.17.39/suite.log.html)

**Comment by mongoose-im** [comment]
[5402.1](https://travis-ci.org/esl/MongooseIM/jobs/418601437) / Erlang 19.3 / small_tests / 6f24bb4d0c9567926b35466904ef8ba68c5744a6
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5402/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5402/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-21_09.09.44/lib.mongooseim.logs/run.2018-08-21_09.09.45/suite.log.html)

File: big_tests/tests/rest_client_SUITE.erl [hunk 6]
```diff
@@ -251,6 +275,45 @@ room_is_created_with_given_identifier(Config) ->
...
+config_can_be_changed_by_owner(Config) ->
+    escalus:fresh_story(Config, [{alice, 1}], fun(Alice) ->
+        RoomID = <<"som_othere_id">>,
+        RoomJID = room_jid(RoomID, Config),
+        RoomID = given_new_room({alice, Alice}, RoomJID, <<"old_name">>),
+        RoomInfo = get_room_info({alice, Alice}, RoomID),
+        assert_property_value(<<"name">>, <<"old_name">>, RoomInfo),
+
+        {{<<"200">>, <<"OK">>}, {_}} =
+            given_config_change({alice, Alice}, RoomJID, <<"new_name">>, <<"new_subject">>),
```
Comment (fenek) [inline big_tests/tests/rest_client_SUITE.erl]: This is the "when" part, not "given".

File: big_tests/tests/rest_client_SUITE.erl [hunk 7]
```diff
@@ -251,6 +275,45 @@ room_is_created_with_given_identifier(Config) ->
...
+        assert_property_value(<<"name">>, <<"new_name">>, NewRoomInfo),
+        assert_property_value(<<"subject">>, <<"new_subject">>, NewRoomInfo)
+    end).
+
+config_cannot_be_changed_by_member(Config) ->
+    escalus:fresh_story(Config, [{alice, 1}, {bob, 1}], fun(Alice, Bob) ->
+        RoomID = given_new_room_with_users({alice, Alice}, [{bob, Bob}]),
+        RoomJID = room_jid(RoomID, Config),
+        {{<<"403">>,<<"Forbidden">>},<<>>} =
+            given_config_change({bob, Bob}, RoomJID, <<"other_name">>, <<"other_subject">>),
```
Comment (fenek) [inline big_tests/tests/rest_client_SUITE.erl]: Same as above.

File: big_tests/tests/rest_client_SUITE.erl [hunk 8]
```diff
@@ -251,6 +275,45 @@ room_is_created_with_given_identifier(Config) ->
...
+    end).
+
+config_can_be_changed_by_all(Config) ->
+    escalus:fresh_story(Config, [{alice, 1}, {bob, 1}], fun(Alice, Bob) ->
+        RoomID = given_new_room_with_users({alice, Alice}, [{bob, Bob}]),
+        RoomJID = room_jid(RoomID, Config),
+        RoomInfo = get_room_info({alice, Alice}, RoomID),
+        assert_property_value(<<"name">>,<<"new_room_name">>,RoomInfo),
+        {{<<"200">>, <<"OK">>}, {_}} =
+            given_config_change({bob, Bob}, RoomJID, <<"other_name">>, <<"other_subject">>),
```
Comment (fenek) [inline big_tests/tests/rest_client_SUITE.erl]: Same as above.

File: big_tests/tests/rest_client_SUITE.erl [hunk 9]
```diff
@@ -1081,3 +1158,10 @@ break_stuff(Config) ->
 room_jid(RoomID, Config) ->
     MUCLightHost = ?config(muc_light_host, Config),
     <<RoomID/binary, "@", MUCLightHost/binary>>.
+
+set_mod_config(K, V, Config) ->
```
Comment (fenek) [inline big_tests/tests/rest_client_SUITE.erl]: Duplicate of the same function from `muc_light_SUITE` and `muc_light_legacy_SUITE`. Might be extracted from all 3 suites to `muc_light_helper`.

File: src/muc_light/mod_muc_light_commands.erl [hunk 10]
```diff
@@ -70,15 +69,16 @@ commands() ->
        ]},
       {result, {name, binary}}],
 
-     [{name, create_identifiable_muc_light_room},
+     [{name, change_config_or_create_identifiable_room},
       {category, <<"muc-lights">>},
-      {desc, <<"Create a MUC Light room with user-provided username part in JID">>},
+      {desc, <<"Specify muc light room configuration and create/update room with the gived id.">>},
```
Comment (fenek) [inline src/muc_light/mod_muc_light_commands.erl]: MUC Light
given

**Comment by mongoose-im** [comment]
[5403.1](https://travis-ci.org/esl/MongooseIM/jobs/418726754) / Erlang 19.3 / small_tests / 6e3fd312a9d6b8d613d170559ec9cb503470464f
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5403/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5403/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-21_14.33.00/lib.mongooseim.logs/run.2018-08-21_14.33.01/suite.log.html)

**Comment by mongoose-im** [comment]
[5404.1](https://travis-ci.org/esl/MongooseIM/jobs/419036969) / Erlang 19.3 / small_tests / fb7cf404190e8a051c8fcbba7100316254a59153
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5404/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5404/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-22_07.06.41/lib.mongooseim.logs/run.2018-08-22_07.06.41/suite.log.html)

**Comment by mongoose-im** [comment]
[5409.1](https://travis-ci.org/esl/MongooseIM/jobs/419164027) / Erlang 19.3 / small_tests / 1746fe804e0fb7fad4e02c43863fca582973e2bf
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5409/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5409/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-22_13.38.57/lib.mongooseim.logs/run.2018-08-22_13.38.57/suite.log.html)

**Comment by mongoose-im** [comment]
[5412.1](https://travis-ci.org/esl/MongooseIM/jobs/419191604) / Erlang 19.3 / small_tests / 0da6e19a392cd5a2cf8c64280d38f535524702ca
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5412/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5412/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-22_14.06.43/lib.mongooseim.logs/run.2018-08-22_14.06.44/suite.log.html)

**Comment by mongoose-im** [comment]
[5413.1](https://travis-ci.org/esl/MongooseIM/jobs/419203305) / Erlang 19.3 / small_tests / eed249df6c52d0658799091f638c486152cacf8a
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5413/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5413/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-22_14.31.24/lib.mongooseim.logs/run.2018-08-22_14.31.25/suite.log.html)

File: rel/files/ejabberd.cfg [hunk 11]
```diff
@@ -198,6 +198,7 @@
           {"_", "/api/messages/[:with]", mongoose_client_api_messages, []},
           {"_", "/api/contacts/[:jid]", mongoose_client_api_contacts, []},
           {"_", "/api/rooms/[:id]",    mongoose_client_api_rooms, []},
+          {"_", "/api/rooms/[:id]/config",    mongoose_client_api_rooms_config, []},
```
[the remark this answers is not in the document]
Comment (zofpolkowska) [reply rel/files/ejabberd.cfg]: added :)

**Comment by mongoose-im** [comment]
[5418.1](https://travis-ci.org/esl/MongooseIM/jobs/419229236) / Erlang 19.3 / small_tests / d8a9c0f53d1ca01551a8cd59af9692abb4449a64
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5418/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5418/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-22_15.36.09/lib.mongooseim.logs/run.2018-08-22_15.36.09/suite.log.html)

**Comment by mongoose-im** [comment]
[5420.1](https://travis-ci.org/esl/MongooseIM/jobs/419245140) / Erlang 19.3 / small_tests / 7ee8e2b8a1d4d11815f64f39baa57ada6113d90c
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5420/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5420/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-22_16.28.45/lib.mongooseim.logs/run.2018-08-22_16.28.46/suite.log.html)

**Comment by mongoose-im** [comment]
[5424.1](https://travis-ci.org/esl/MongooseIM/jobs/419513349) / Erlang 19.3 / small_tests / d09fbd81e153cca9c9a0424043a7a0e0ec4a5830
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5424/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5424/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-23_06.51.18/lib.mongooseim.logs/run.2018-08-23_06.51.18/suite.log.html)

**Comment by mongoose-im** [comment]
[5425.1](https://travis-ci.org/esl/MongooseIM/jobs/419522930) / Erlang 19.3 / small_tests / 8f57b66534a343bfc481d85a38dc4c0c4565f0f9
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2030/5425/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2030/5425/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-08-23_07.26.37/lib.mongooseim.logs/run.2018-08-23_07.26.38/suite.log.html)

**Pull request merged**